edx_when/api.py

```diff
@@ -357,6 +363,47 @@ def set_date_for_block(course_id, block_id, field, date_or_timedelta, user=None,
         return existing_date.id
 
 
+def get_schedules_with_due_date(course_id, assignment_date):
+    """
+    Get all Schedules with assignments due on a specific date for a Course.
+
+    Arguments:
+        course_id: either a CourseKey or string representation of same
+        assignment_date: a date object
+
+    Returns:
+        a QuerySet of Schedule objects for Users who have content due on the specified assignment_date
+    """
+    user_ids = models.UserDate.objects.select_related('content_date').select_related('content_date__policy').annotate(
+        computed_date=ExpressionWrapper(
+            F('content_date__policy__abs_date') + F('rel_date'),
+            output_field=DateTimeField()
+        ),
+    ).filter(
+        Q(computed_date__date=assignment_date, abs_date__isnull=True) |
+        Q(content_date__policy__abs_date__date=assignment_date, rel_date__isnull=True)
+    ).values_list('user_id', flat=True).distinct()
+
+    user_date_overridden_schedules = Schedule.objects.filter(
+        enrollment__course_id=course_id,
+        enrollment__user_id__in=user_ids,
+    )
+
+    # Get all relative dates for a course, we want them distinct, it doesn't matter how many of each due date there is
+    rel_dates = models.ContentDate.objects.filter(
+        course_id=course_id,
+    ).select_related('policy').values_list('policy__rel_date', flat=True).distinct()
```

Comment:
When you use `values_list`, do you get a subquery `WHERE foo IN (SELECT * FROM ...)` or a list query `WHERE foo in (1, 2, 3, ...)`? I think the former is going to perform better, and you can get it by using `values('policy__rel_date').distinct()`

Reply:
The goal being to not actually pull any data into python until the final query.

Reply:
Double checked and I'll post the rest of the queries, but the relevant resulting sql is ```    "test_models_app_dummyenrollment"."user_id" IN (
        SELECT DISTINCT U0."user_id" 
        FROM "edx_when_userdate" U0 
        INNER JOIN "edx_when_contentdate" U1 
        ON (U0."content_date_id" = U1."id") 
        INNER JOIN "edx_when_datepolicy" U2 
        ON (U1."policy_id" = U2."id") 
        WHERE (
            (
                U0."abs_date" IS NULL AND 
                django_datetime_cast_date((django_format_dtdelta('+', U2."abs_date", U0."rel_date")), NULL) = 2019-03-22
            ) OR (
                django_datetime_cast_date(U2."abs_date", NULL) = 2019-03-22 AND U0."rel_date" IS NULL
            )
        )
    )```

so I think we're good to go as it does a subquery.

Reply:
Yup, looks good.